File: packages/digital-ink-recognition/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/digitalinkrecognition/DigitalInkRecognitionPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.CustomException;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.options.DeleteDownloadedModelOptions;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.options.DownloadModelOptions;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.options.RecognizeOptions;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.results.GetDownloadedModelsResult;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.results.RecognizeResult;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.interfaces.Result;

@CapacitorPlugin(name = "DigitalInkRecognition")
public class DigitalInkRecognitionPlugin extends Plugin {

    public static final String TAG = "DigitalInkRecognition";

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error occurred.";

    private DigitalInkRecognition implementation;

    @Override
    public void load() {
        implementation = new DigitalInkRecognition();
    }

    @PluginMethod
    public void deleteDownloadedModel(PluginCall call) {
        try {
            DeleteDownloadedModelOptions options = new DeleteDownloadedModelOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.deleteDownloadedModel(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void downloadModel(PluginCall call) {
        try {
            DownloadModelOptions options = new DownloadModelOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.downloadModel(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getDownloadedModels(PluginCall call) {
        try {
            NonEmptyResultCallback<GetDownloadedModelsResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull GetDownloadedModelsResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.getDownloadedModels(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void recognize(PluginCall call) {
        try {
            RecognizeOptions options = new RecognizeOptions(call);
            NonEmptyResultCallback<RecognizeResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull RecognizeResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.recognize(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        if (exception instanceof CustomException) {
            call.reject(message, ((CustomException) exception).getCode());
        } else {
            call.reject(message);
        }
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}

File: packages/digital-ink-recognition/ios/Plugin/DigitalInkRecognitionPlugin.swift
import Foundation
import Capacitor
import MLKitDigitalInkRecognition

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(DigitalInkRecognitionPlugin)
public class DigitalInkRecognitionPlugin: CAPPlugin {
    public static let tag = "DigitalInkRecognition"

    private var implementation: DigitalInkRecognition?

    override public func load() {
        implementation = DigitalInkRecognition(plugin: self)

        NotificationCenter.default.addObserver(
            forName: .mlkitModelDownloadDidSucceed,
            object: nil,
            queue: nil
        ) { [weak self] notification in
            self?.handleModelDownloadDidSucceed(notification: notification)
        }

        NotificationCenter.default.addObserver(
            forName: .mlkitModelDownloadDidFail,
            object: nil,
            queue: nil
        ) { [weak self] notification in
            self?.handleModelDownloadDidFail(notification: notification)
        }
    }

    @objc func deleteDownloadedModel(_ call: CAPPluginCall) {
        do {
            let options = try DeleteDownloadedModelOptions(call)

            try implementation?.deleteDownloadedModel(options, completion: { error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func downloadModel(_ call: CAPPluginCall) {
        do {
            let options = try DownloadModelOptions(call)

            try implementation?.downloadModel(options, completion: { error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func getDownloadedModels(_ call: CAPPluginCall) {
        implementation?.getDownloadedModels(completion: { result, error in
            if let error = error {
                self.rejectCall(call, error)
            } else {
                self.resolveCall(call, result)
            }
        })
    }

    @objc func recognize(_ call: CAPPluginCall) {
        do {
            let options = try RecognizeOptions(call)

            try implementation?.recognize(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call, result)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    private func handleModelDownloadDidFail(notification: Notification) {
        guard let userInfo = notification.userInfo,
              let model = userInfo[ModelDownloadUserInfoKey.remoteModel.rawValue]
                as? DigitalInkRecognitionModel else {
            return
        }
        let error = userInfo[ModelDownloadUserInfoKey.error.rawValue] as? Error
        implementation?.handleModelDownloadDidFail(model: model, error: error)
    }

    private func handleModelDownloadDidSucceed(notification: Notification) {
        guard let userInfo = notification.userInfo,
              let model = userInfo[ModelDownloadUserInfoKey.remoteModel.rawValue]
                as? DigitalInkRecognitionModel else {
            return
        }
        implementation?.handleModelDownloadDidSucceed(model: model)
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", DigitalInkRecognitionPlugin.tag, "] ", error)
        if let customError = error as? CustomError {
            call.reject(error.localizedDescription, customError.code)
        } else {
            call.reject(error.localizedDescription)
        }
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
